File: src/main/java/com/myapp/controller/AuthController.java
package com.myapp.controller;

import com.myapp.auth.TokenHandler;
import com.myapp.service.SecurityContextService;
import lombok.Value;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.security.authentication.AuthenticationManager;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.AuthenticationException;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestMethod;
import org.springframework.web.bind.annotation.RestController;

/**
 * @author riccardo.causo
 */
@RestController
@RequestMapping("/api/auth")
public class AuthController {

    private final AuthenticationManager authenticationManager;
    private final TokenHandler tokenHandler;
    private final SecurityContextService securityContextService;

    @Autowired
    AuthController(AuthenticationManager authenticationManager,
                   TokenHandler tokenHandler,
                   SecurityContextService securityContextService) {
        this.authenticationManager = authenticationManager;
        this.tokenHandler = tokenHandler;
        this.securityContextService = securityContextService;
    }

    @RequestMapping(method = RequestMethod.POST)
    public AuthResponse auth(@RequestBody AuthParams params) throws AuthenticationException {
        final UsernamePasswordAuthenticationToken loginToken = params.toAuthenticationToken();
        final Authentication authentication = authenticationManager.authenticate(loginToken);
        SecurityContextHolder.getContext().setAuthentication(authentication);

        return securityContextService.currentUser().map(u -> {
            final String token = tokenHandler.createTokenForUser(u);
            return new AuthResponse(token);
        }).orElseThrow(RuntimeException::new); // it does not happen.
    }

    @Value
    private static final class AuthParams {
        private final String email;
        private final String password;

        UsernamePasswordAuthenticationToken toAuthenticationToken() {
            return new UsernamePasswordAuthenticationToken(email, password);
        }
    }

    @Value
    private static final class AuthResponse {
        private final String token;
    }

}

File: src/main/kotlin/com/myapp/controller/AuthController.kt
package com.myapp.controller

import com.myapp.auth.TokenHandler
import com.myapp.auth.SecurityContextService
import org.slf4j.LoggerFactory
import org.springframework.security.authentication.AuthenticationManager
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken
import org.springframework.security.core.context.SecurityContextHolder
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController


@RestController
@RequestMapping("/api/auth")
class AuthController(
    private val authenticationManager: AuthenticationManager,
    private val tokenHandler: TokenHandler,
    private val securityContextService: SecurityContextService
) {

    @Suppress("unused")
    private val logger = LoggerFactory.getLogger(AuthController::class.java)

    @PostMapping
    fun auth(@RequestBody params: AuthParams): AuthResponse {
        val loginToken = UsernamePasswordAuthenticationToken(params.email, params.password)
        val authentication = authenticationManager.authenticate(loginToken)
        SecurityContextHolder.getContext().authentication = authentication

        return securityContextService.currentUser()
            .let { requireNotNull(it) }
            .let { tokenHandler.createTokenForUser(it).let(::AuthResponse) }
    }

    data class AuthParams(
        val email: String,
        val password: String
    )

    data class AuthResponse(val token: String)

}
